packages/cron/src/jobs/pins-backup.js

```diff
@@ -0,0 +1,239 @@
+import formatNumber from 'format-number'
+import batch from 'it-batch'
+import { pipe } from 'it-pipe'
+import { CID } from 'multiformats'
+import * as raw from 'multiformats/codecs/raw'
+import { Readable } from 'stream'
+
+export default class Backup {
+  constructor (env) {
+    this.UPDATE_BACKUP_URL = `
+      UPDATE psa_pin_request 
+      SET backup_urls=$1
+      WHERE id = $2 AND
+      content_cid= $3
+      RETURNING *
+    `
+    this.fmt = formatNumber()
+    this.SIZE_TIMEOUT = 1000 * 10 // timeout if we can't figure out the size in 10s
+    this.BLOCK_TIMEOUT = 1000 * 30 // timeout if we don't receive a block after 30s
+    this.REPORT_INTERVAL = 1000 * 60 // log download progress every minute
+    this.MAX_DAG_SIZE = 1024 * 1024 * 1024 * 32 // don't try to transfer a DAG that's bigger than 32GB
+    this.log = debug('backup:pins')
+    this.env = env
+    this.LIMIT = env.QUERY_LIMIT !== undefined ? env.QUERY_LIMIT : 10000
+    this.GET_PINNED_PINS_QUERY = `
+      SELECT psa.id, psa.backup_urls, psa.source_cid, psa.content_cid, psa.name, psa.auth_key_id
+      FROM psa_pin_request psa
+        JOIN pin p ON p.content_cid = psa.content_cid
+      WHERE p.status = 'Pinned'
+        AND psa.backup_urls IS NULL
+      LIMIT $1
+    `
+    this.s3 = new AWS.S3({})
+    this.CONCURRENCY = env.CONCURRENCY !== undefined ? env.CONCURRENCY : 10
+  }
+
+  /**
+   * @param {import('pg').Client} db
+   */
+  registerBackup (db, contentCid, pinRequestId) {
+    /**
+     * @param {AsyncIterable<import('./bindings').RemoteBackup} source
+     */
+    return async (source) => {
+      for await (const bak of source) {
+        this.log(`backing up ${JSON.stringify(bak)}`)
+        const res = await db.query(this.UPDATE_BACKUP_URL, [
+          [bak.backupUrl.toString()],
+          pinRequestId,
+          contentCid.toString()
+        ])
+        this.log(`Result from updating pin request ${JSON.stringify(res)}`)
+        this.log(`saved backup record for upload ${bak.contentCid}: ${bak.backupUrl.toString()}, rowId: ${pinRequestId}`)
+      }
+    }
+  }
+
+  /**
+   * @param {import('@aws-sdk/client-s3').S3Client} s3
+   * @param {string} bucketName
+   */
+  uploadCar (s3, bucketName) {
+    /**
+     * @param {AsyncIterable<import('./bindings').BackupContent} source
+     */
+    return async function * (source) {
+      for await (const bak of source) {
+        const backupUrl = await this.s3Upload(s3, bucketName, bak)
+        /** @type {import('./bindings').RemoteBackup} */
+        const backup = { ...bak, backupUrl }
+        yield backup
+      }
+    }
+  }
+
+  /**
+   * @param {import('@aws-sdk/client-s3').S3Client} s3
+   * @param {string} bucketName
+   * @param {import('./bindings').BackupContent} bak
+   */
+  async s3Upload (s3, bucketName, bak) {
+    const key = `complete/${bak.contentCid}.car`
+    const region = await s3.config.region()
+    const url = new URL(`https://${bucketName}.s3.${region}.amazonaws.com/${key}`)
```

Comment:
If I'm not mistaken, with the current implementation, given we can have different PSA_PIN_REQUESTS pointing to the same `CID`, we would end up uploading the same car multiple time to S3 isn't it?
I'm not quite aware if we need a particular folder structre, but assuming this one the correct one (and the same cid would always endup in the same path) we should check if already exists, and if yes, avoid uploading it?

Also, it seems from here that we always use content_cid (and not source_cid) for the key meaning different cid versions are not something we should be caring about?

Reply:
Ok so, I've updated it so it's `complete/${bak.contentCid}/${bak.sourceCid}.car` for the key
And then added it so if that key already exists then we won't re-upload it.

Reply:
Ace!

Reply:
The current convention is `complete/${bak.sourceCid}.car`